src/hooks/sounds.tsx

```diff
@@ -0,0 +1,164 @@
+import React, {
+  createContext,
+  useCallback,
+  useContext,
+  useEffect,
+  useMemo,
+  useState,
+} from "react";
+
+import useSound from "use-sound";
+import {
+  ExposedData,
+  HookOptions,
+  PlayFunction,
+  PlayOptions,
+} from "use-sound/dist/types";
+import { HowlOptions } from "howler";
+
+import { SoundConfigMap, SoundConfigReference } from "types/sounds";
+
+import { fetchSoundConfigs } from "api/sounds";
+
+export type PlaySpriteFunction = (options?: PlaySpriteOptions) => void;
+export type PlaySpriteOptions = Omit<PlayOptions, "id">;
+
+export type UseCustomSoundOptions = HookOptions & HowlOptions;
+
+export interface ExposedDataWithPlay extends ExposedData {
+  play: PlayFunction;
+}
+
+export interface CustomSoundsState {
+  soundConfigs: SoundConfigMap;
+}
+
+export const initialValue: CustomSoundsState = {
+  soundConfigs: {},
+};
+
+export const CustomSoundsContext = createContext<CustomSoundsState>(
+  initialValue
+);
+
+export const CustomSoundsProvider: React.FC = ({ children }) => {
+  const [soundConfigs, setSoundConfigs] = useState<SoundConfigMap>(
+    initialValue.soundConfigs
+  );
+
+  // Fetch the sound configs data on first load
+  useEffect(() => {
+    fetchSoundConfigs().then((soundConfigs) => {
+      setSoundConfigs(soundConfigs);
+    });
+  }, []);
+
+  const providerData = useMemo(
+    () => ({
+      soundConfigs,
+    }),
+    [soundConfigs]
+  );
+
+  return (
+    <CustomSoundsContext.Provider value={providerData}>
+      {children}
+    </CustomSoundsContext.Provider>
+  );
+};
```

Comment:
Not against it, but just want to understand. What is the benefit of doing it this way, instead of our `default` useFirestoreConnect?. Even though I do understand that we only need to fetch it one time, it seems to be very verbose, compared to useFirestoreConnect

Reply:
We want to move away from useFirestoreConnect as it’s more or less abandoned.

This context pattern means that we are actually literally only calling the code once, and providing that where it’s needed. This is the singleton pattern I described that we should have used for a lot of our current ‘needs to be called once’ hooks. I would say a lot of our current ‘use once’ hooks only sort of work by crude side effect rather than good patterns (because useFirestoreConnect /the firestore SDK attempts to de-duper/reuse configs/caches/etc so as not to make extra unnneded connections; but we’ve seen bugs in this before)

Context’s are a very common, standard, and recommended way of building clean react code (despite the opinions of one of our former engineers who ripped some of the previous ones out)

The ‘verbosity’ of the pattern is a few extra lines of boilerplate during setup, and is on par with the equivalent requirements of a redux setup (eg initialState, etc)

The other alternative would be to move it all to redux, which would have just as much boilerplate (if not more), and would still require us to ‘connect it’ at some point, which would end up looking very similar. I expect at the end of the day the difference between implementation would be that the data is stored in redux and we use a selector to retrieve it, vs us using a useState and already having the sata ‘selected’.

Reply:
Also, if this is a pattern we start to use more, I can absolutely abstract this boilerplate away into a nicer 'API' using helpers/etc; but given the urgency of the timeline around both the loaders and the sounds features, I decided it wasn't worth 'refining' to that level at this stage, particularly as it would likely only save a few lines of boilerplate anyway.